### PR TITLE
ODB support dbGuide

### PR DESCRIPTION
**Pull request by osamahammad21**

### CODE REVIEW

**Comment by osamahammad21** [comment]
@maliberty I am not sure how to implement route guides in gui since drawing all of them would cover the whole design and would be messy.

**Comment by maliberty** [comment]
I suggest you implement an inspector so that when a net is selected you can choose to see it's guides.

You could make a gcell inspector as well showing all the nets through the gcell though that might be slow to generate

File: src/gui/src/dbDescriptors.cpp [hunk 1]
```diff
@@ -1311,7 +1311,11 @@ Descriptor::Actions DbNetDescriptor::getActions(std::any object) const
       return makeSelected(net, nullptr);
     }});
   }
-
+  if(!net->getGuides().empty())
+    actions.push_back(Descriptor::Action{"Route Guides", [this, gui, net]() {
+      gui->addRouteGuides(net);
+      return makeSelected(net, nullptr);
+    }});
```
Comment (arlpetergadfort) [inline src/gui/src/dbDescriptors.cpp:1318]: There should be a way to remove the route guide as well.

File: src/gui/src/layoutViewer.cpp [hunk 2]
```diff
@@ -3032,6 +3037,14 @@ void LayoutViewer::addFocusNet(odb::dbNet* net)
   }
 }
 
+void LayoutViewer::addRouteGuides(odb::dbNet* net)
+{
+  const auto& [itr, inserted] = route_guides_.insert(net);
+  if (inserted) {
+    fullRepaint();
+  }
```
Comment (arlpetergadfort) [inline src/gui/src/layoutViewer.cpp:3045]: This should be handled in the renderer you added, otherwise there is no point in having it.

File: src/gui/src/layoutViewer.cpp [hunk 3]
```diff
@@ -3049,6 +3062,14 @@ void LayoutViewer::clearFocusNets()
...
 }
 
+void LayoutViewer::clearRouteGuides()
+{
+  if (!route_guides_.empty()) {
+    route_guides_.clear();
+    fullRepaint();
+  }
+}
+
```
Comment (arlpetergadfort) [inline src/gui/src/layoutViewer.cpp:3072]: Same

File: src/gui/src/layoutViewer.h [hunk 4]
```diff
@@ -423,6 +427,8 @@ class LayoutViewer : public QWidget
 
   // Set of nets to focus drawing on, if empty draw everything
   std::set<odb::dbNet*> focus_nets_;
+  // Set of nets to draw route guides for, if empty draw nothing
+  std::set<odb::dbNet*> route_guides_;
```
Comment (arlpetergadfort) [inline src/gui/src/layoutViewer.h:431]: In guide renderer.

File: src/gui/src/layoutViewer.h [hunk 5]
```diff
@@ -136,8 +137,11 @@ class LayoutViewer : public QWidget
 
   void addFocusNet(odb::dbNet* net);
   void removeFocusNet(odb::dbNet* net);
+  void addRouteGuides(odb::dbNet* net);
```
Comment (arlpetergadfort) [inline src/gui/src/layoutViewer.h:140]: In renderer.

File: src/gui/src/layoutViewer.h [hunk 6]
```diff
@@ -136,8 +137,11 @@ class LayoutViewer : public QWidget
...
   void clearFocusNets();
+  void clearRouteGuides();
   const std::set<odb::dbNet*>& getFocusNets() { return focus_nets_; }
+  const std::set<odb::dbNet*>& getRouteGuides() { return route_guides_; }
```
Comment (arlpetergadfort) [inline src/gui/src/layoutViewer.h:144]: In renderer.

File: src/gui/src/mainWindow.cpp [hunk 7]
```diff
@@ -400,9 +400,11 @@ void MainWindow::init(sta::dbSta* sta)
   timing_widget_->init(sta);
   controls_->setSTA(sta);
   hierarchy_widget_->setSTA(sta);
-
-  // register descriptors
+  // register renderers
   auto* gui = Gui::get();
+  guide_renderer_ = std::make_unique<GuideRenderer>(viewer_->getRouteGuides());
+  gui->registerRenderer(guide_renderer_.get());
```
Comment (arlpetergadfort) [inline src/gui/src/mainWindow.cpp:406]: DRT or GRT could hold on to this renderer instead of making it part of the main_window. We can figure out a way to pass the actions along to the layout viewer

File: src/gui/src/gui.cpp [hunk 8]
```diff
@@ -856,6 +856,11 @@ void Gui::addFocusNet(odb::dbNet* net)
   main_window->getLayoutViewer()->addFocusNet(net);
 }
 
+void Gui::addRouteGuides(odb::dbNet* net)
+{
+  main_window->getLayoutViewer()->addRouteGuides(net);
+}
```
Comment (arlpetergadfort) [inline src/gui/src/gui.cpp:862]: DRT or GRT should probably have these. and Only register the renderer when something is added and remove it when it's empty.

File: src/gui/src/layoutViewer.cpp [hunk 9]
```diff
@@ -2921,6 +2921,7 @@ void LayoutViewer::addMenuAndActions()
   menu_actions_[CLEAR_HIGHLIGHTS_ACT] = clear_menu->addAction(tr("Highlights"));
   menu_actions_[CLEAR_RULERS_ACT] = clear_menu->addAction(tr("Rulers"));
   menu_actions_[CLEAR_FOCUS_ACT] = clear_menu->addAction(tr("Focus nets"));
+  menu_actions_[CLEAR_GUIDES_ACT] = clear_menu->addAction(tr("Route Guides"));
```
Comment (arlpetergadfort) [inline src/gui/src/layoutViewer.cpp:2924]: I think this is okay for now, but we probably need a better way to add these for the renderers so the code doesn't get too much more co-mingled.

File: src/gui/src/dbDescriptors.cpp [hunk 10]
```diff
@@ -1313,7 +1315,10 @@ Descriptor::Actions DbNetDescriptor::getActions(std::any object) const
   }
   if(!net->getGuides().empty())
     actions.push_back(Descriptor::Action{"Route Guides", [this, gui, net]() {
-      gui->addRouteGuides(net);
+      if(guide_nets_.count(net) == 0)
+        gui->addRouteGuides(net);
+      else
+        gui->removeRouteGuides(net);
```
Comment (arlpetergadfort) [inline src/gui/src/dbDescriptors.cpp:1321]: Take a look at the above code for the Focus and De-focus. I think it might be better to have an action added for Route Guide and one for Hide the route guide, so it's clear what the action does.

Comment (osamahammad21) [reply src/gui/src/dbDescriptors.cpp:1321]: I agree with you but I think this can be an improvement on gui that can be done in another PR. I think for now it serves the purpose of debugging the route guides on the gui.

Comment (arlpetergadfort) [reply src/gui/src/dbDescriptors.cpp:1321]: I'd be okay with it being done in another PR.

File: src/gui/src/layoutViewer.cpp [hunk 11]
```diff
@@ -2141,6 +2141,20 @@ void LayoutViewer::drawRegionOutlines(QPainter* painter)
...
+void LayoutViewer::drawRouteGuides(Painter& painter, odb::dbTechLayer* layer)
+{
+  if(route_guides_.empty())
+    return;
+  for(auto net : route_guides_) {
+    for(auto guide : net->getGuides()) {
+      if (guide->getLayer() != layer)
+        continue;
+      painter.setPen(layer);
+      painter.setBrush(layer);
```
[the remark this answers is not in the document]
Comment (osamahammad21) [reply src/gui/src/layoutViewer.cpp]: Done

File: src/drt/include/triton_route/TritonRoute.h [hunk 12]
```diff
@@ -69,8 +69,6 @@ namespace triton_route {
 
 struct ParamStruct
 {
-  std::string guideFile;
```
Comment (maliberty) [inline src/drt/include/triton_route/TritonRoute.h:72]: The ISPD benchmarks require us to read a guide file not generated by OR.  Is that still possible?

File: src/drt/test/ispd18_sample.defok [hunk 13]
```diff
@@ -27,6 +27,8 @@ TRACKS X 83800 DO 52 STEP 400 LAYER Metal2 ;
 TRACKS Y 72010 DO 51 STEP 380 LAYER Metal2 ;
 TRACKS X 83800 DO 52 STEP 400 LAYER Metal1 ;
 TRACKS Y 72010 DO 51 STEP 380 LAYER Metal1 ;
+GCELLGRID X 83600 DO 18 STEP 5700 ;
+GCELLGRID Y 71820 DO 16 STEP 5700 ;
```
Comment (maliberty) [inline src/drt/test/ispd18_sample.defok:31]: Why is this now necessary?

File: src/drt/include/triton_route/TritonRoute.h (hunk 12 above)
Comment (osamahammad21) [reply src/drt/include/triton_route/TritonRoute.h:72]: Already handled, check run-ispd.py

Comment (maliberty) [reply src/drt/include/triton_route/TritonRoute.h:72]: Nevermind I see you have a read_guides command.

File: src/odb/src/db/dbBlock.cpp [hunk 14]
```diff
@@ -3292,6 +3293,38 @@ void dbBlock::writeDb(char* filename, int allNode)
   }
 }
 
+void dbBlock::writeGuides(const char* filename) const
+{
+  std::ofstream guide_file;
+  guide_file.open(filename);
+  if (!guide_file.is_open()) {
```
Comment (maliberty) [inline src/odb/src/db/dbBlock.cpp:3300]: if (!guide_file) {

File: src/odb/src/db/dbNet.cpp [hunk 15]
```diff
@@ -95,14 +96,15 @@ _dbNet::_dbNet(_dbDatabase*, const _dbNet& n)
       _ccAdjustOrder(n._ccAdjustOrder)
 
 {
+  guide_tbl_ = new dbTable<_dbGuide>(db, this, *n.guide_tbl_);
```
Comment (maliberty) [inline src/odb/src/db/dbNet.cpp:99]: A table per net is going to be costly in terms of memory.  The table is designed for a lot of objects and allocates in chunks while a single net may not have many guides.

The table should be on the block and the guide can have prev/next-in-net id field for a linked list of guides per net.  The net would have a head guide id.  Similar to bterms on the net.

File: src/odb/src/db/odb.tcl [hunk 16]
```diff
@@ -508,6 +508,20 @@ proc report_group { group } {
...
 
+sta::define_cmd_args "write_guides" { filename }
+
+proc write_guides { args } {
+  sta::check_argc_eq1 "write_guides" $args
+  set filename $args
+  set db [ord::get_db]
+  set chip [$db getChip]
+  if { $chip == "NULL" } {
+    utl::error "please load the design before trying to use this command"
```
Comment (maliberty) [inline src/odb/src/db/odb.tcl:519]: This needs ODB and an id

File: src/drt/src/io/io_parser_helper.cpp [hunk 17]
```diff
@@ -714,16 +716,12 @@ void io::Parser::postProcessGuide(odb::dbDatabase* db)
   logger->info(DRT, 179, "Init gr pin query.");
   design->getRegionQuery()->initGRPin(tmpGRPins);
 
-  if (OUTGUIDE_FILE == string("")) {
+  if (!SAVE_GUIDE_UPDATES) {
     if (VERBOSE > 0) {
-      logger->warn(
-          DRT, 245, "No output guide specified, skipped writing guide.");
+      logger->warn(DRT, 245, "skipped writing guide updates to database.");
```
Comment (maliberty) [inline src/drt/src/io/io_parser_helper.cpp:721]: info seems enough as nothing has gone wrong

File: src/grt/test/est_rc1.ok [hunk 18]
```diff
@@ -69,6 +69,7 @@ metal10           2209             0            0.00%             0 /  0 /  0
 Total           117668          3386            2.88%             0 /  0 /  0
 
 [INFO GRT-0018] Total wirelength: 10336 um
+[INFO GRT-0014] Routed nets: 563
```
Comment (maliberty) [inline src/grt/test/est_rc1.ok:72]: Why is there a new message here?  It seems unrelated

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/drt/test/ispd18_sample.defok (hunk 13 above)
Comment (osamahammad21) [reply src/drt/test/ispd18_sample.defok:31]: Because now instead of reading guides only by tritonroute, we read guides by fastroute as well which initializes the gcell grid from the given guides.

File: src/odb/src/db/odb.tcl (hunk 16 above)
Comment (maliberty) [reply src/odb/src/db/odb.tcl:519]: please do

File: src/grt/test/est_rc1.ok (hunk 18 above)
Comment (maliberty) [reply src/grt/test/est_rc1.ok:72]: reading the guides generates a message about the number of routed nets?

File: src/drt/test/single_step.ok [hunk 19]
```diff
@@ -62,9 +62,9 @@
 [INFO DRT-0036] FR_VIA guide region query size = 0.
 [INFO DRT-0036] Metal1 guide region query size = 22.
 [INFO DRT-0036] Via1 guide region query size = 0.
-[INFO DRT-0036] Metal2 guide region query size = 22.
+[INFO DRT-0036] Metal2 guide region query size = 21.
 [INFO DRT-0036] Via2 guide region query size = 0.
-[INFO DRT-0036] Metal3 guide region query size = 10.
+[INFO DRT-0036] Metal3 guide region query size = 15.
```
[the remark this answers is not in the document]
Comment (osamahammad21) [reply src/drt/test/single_step.ok:67]: I guess because the gcell grid is initialized differently by GRT that it reduces the guide batches needed.

**Review by maliberty — APPROVED** [review]
